File: Core/src/main/kotlin/jokatu/components/ui/DialogManager.java
package jokatu.components.ui;

import jokatu.components.stomp.StoringMessageSender;
import jokatu.game.Game;
import jokatu.game.GameID;
import jokatu.game.exception.GameException;
import jokatu.game.player.Player;
import jokatu.ui.Dialog;
import ophelia.collections.list.UnmodifiableList;
import ophelia.function.ExceptionalConsumer;
import ophelia.tuple.Pair;
import ophelia.util.MapUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Supplier;

import static java.util.Collections.emptyList;

@Component
public class DialogManager implements DialogRequestor, DialogResponder {

	private final StoringMessageSender sender;

	private final Map<DialogID, ExceptionalConsumer<Map<String, Object>, GameException>> dialogs = new ConcurrentHashMap<>();

	private final Map<Pair<GameID, String>, List<DialogUI>> playersDialogs = new ConcurrentHashMap<>();

	private final Supplier<String> idSupplier = new Supplier<String>() {
		private final AtomicLong id = new AtomicLong();

		@NotNull
		@Override
		public String get() {
			return Long.toHexString(id.getAndIncrement());
		}
	};

	@Autowired
	public DialogManager(StoringMessageSender sender) {
		this.sender = sender;
	}

	@Override
	public void requestDialog(
			@NotNull final Dialog dialog,
			@NotNull String playerName,
			@NotNull GameID gameId,
			@NotNull ExceptionalConsumer<Map<String, Object>, GameException> jsonConsumer
	) {
		DialogID dialogId = generateId(gameId);
		Pair<GameID, String> gamePlayerKey = new Pair<>(gameId, playerName);

		DialogUI dialogUi = new DialogUI(dialog, dialogId.getDialogId());

		MapUtils.updateListBasedMap(playersDialogs, gamePlayerKey, dialogUi);

		updatePlayerDialogs(gamePlayerKey);

		dialogs.put(dialogId, jsonConsumer);
	}

	private void updatePlayerDialogs(@NotNull Pair<GameID, String> gamePlayerKey) {
		List<DialogUI> dialogUIs = playersDialogs.get(gamePlayerKey);
		assert dialogUIs != null;

		sender.sendToUser(gamePlayerKey.getRight(), "/topic/dialogs.game." + gamePlayerKey.getLeft(), dialogUIs);
	}

	@NotNull
	private DialogID generateId(@NotNull GameID gameId) {
		return new DialogID(gameId, idSupplier.get());
	}

	@Override
	public void respondToDialog(
			@NotNull DialogID dialogId,
			@NotNull String playerName,
			@NotNull Map<String, Object> json
	) throws GameException {
		Pair<GameID, String> gamePlayerKey = new Pair<>(dialogId.getGameId(), playerName);
		boolean isActuallyPlayersDialog = playersDialogs.getOrDefault(gamePlayerKey, emptyList())
				.stream()
				.anyMatch(ui -> ui.getDialogId().equals(dialogId.getDialogId()));

		ExceptionalConsumer<Map<String, Object>, GameException> dialog = dialogs.get(dialogId);
		if (dialog == null || !isActuallyPlayersDialog) {
			throw new GameException(
					dialogId.getGameId(),
					"Dialog with ID ''{0}'' could not be found.",
					dialogId.getDialogId()
			);
		}
		dialog.accept(json);

		List<DialogUI> dialogUIs = playersDialogs.get(gamePlayerKey);
		if (dialogUIs == null) {
			throw new GameException(
					dialogId.getGameId(),
					"You have no dialogs.",
					dialogId.getDialogId()
			);
		}
		dialogUIs.removeIf(dialogUI -> dialogUI.getDialogId().equals(dialogId.getDialogId()));
		updatePlayerDialogs(gamePlayerKey);
	}

	public <P extends Player> UnmodifiableList<DialogUI> getDialogsForPlayer(
			@NotNull final Game<P> game,
			@NotNull final P player
	) {
		return new UnmodifiableList<>(playersDialogs.get(new Pair<>(game.getIdentifier(), player.getName())));
	}

	public static class DialogUI extends Dialog {
		private final String dialogId;

		private DialogUI(@NotNull Dialog dialog, String dialogId) {
			super(dialog.getTitle(), dialog.getMessage(), dialog.getForm(), dialog.isCancellable());
			this.dialogId = dialogId;
		}

		@NotNull
		public String getDialogId() {
			return dialogId;
		}
	}
}

File: Core/src/main/kotlin/jokatu/components/ui/DialogManager.kt
package jokatu.components.ui

import jokatu.components.stomp.StoringMessageSender
import jokatu.game.Game
import jokatu.game.GameID
import jokatu.game.exception.GameException
import jokatu.game.player.Player
import jokatu.ui.Dialog
import ophelia.collections.list.UnmodifiableList
import ophelia.function.ExceptionalConsumer
import ophelia.tuple.Pair
import ophelia.util.MapUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.Collections.emptyList
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Supplier

@Component
class DialogManager
@Autowired constructor(
		private val sender: StoringMessageSender
) : DialogRequestor, DialogResponder {

	private val dialogs = ConcurrentHashMap<DialogResponder.DialogID, ExceptionalConsumer<Map<String, Any>, GameException>>()

	private val playersDialogs = ConcurrentHashMap<Pair<GameID, String>, MutableList<DialogUI>>()

	private val idSupplier = object : Supplier<String> {
		private val id = AtomicLong()

		override fun get(): String {
			return java.lang.Long.toHexString(id.getAndIncrement())
		}
	}

	override fun requestDialog(
			dialog: Dialog,
			playerName: String,
			gameId: GameID,
			jsonConsumer: ExceptionalConsumer<Map<String, Any>, GameException>
	) {
		log.debug("Requesting dialog $dialog for $playerName in game $gameId")
		val dialogId = generateId(gameId)
		val gamePlayerKey = Pair(gameId, playerName)

		val dialogUi = DialogUI(dialog, dialogId.dialogId)

		MapUtils.updateListBasedMap(playersDialogs, gamePlayerKey, dialogUi)

		updatePlayerDialogs(gamePlayerKey)

		dialogs[dialogId] = jsonConsumer
	}

	private fun updatePlayerDialogs(gamePlayerKey: Pair<GameID, String>) {
		val dialogUIs = playersDialogs[gamePlayerKey]!!

		sender.sendToUser(gamePlayerKey.right, "/topic/dialogs.game." + gamePlayerKey.left, dialogUIs)
	}

	private fun generateId(gameId: GameID): DialogResponder.DialogID {
		return DialogResponder.DialogID(gameId, idSupplier.get())
	}

	@Throws(GameException::class)
	override fun respondToDialog(
			dialogId: DialogResponder.DialogID,
			playerName: String,
			json: Map<String, Any>
	) {
		val gamePlayerKey = Pair(dialogId.gameId, playerName)
		val isActuallyPlayersDialog = playersDialogs.getOrDefault(gamePlayerKey, emptyList())
				.stream()
				.anyMatch { ui -> ui.dialogId == dialogId.dialogId }

		val dialog = dialogs[dialogId]
		if (dialog == null || !isActuallyPlayersDialog) {
			throw GameException(
					dialogId.gameId,
					"Dialog with ID '${dialogId.dialogId}' could not be found."
			)
		}
		dialog.accept(json)

		val dialogUIs = playersDialogs[gamePlayerKey] ?: throw GameException(
				dialogId.gameId,
				"You have no dialogs."
		)
		dialogUIs.removeIf { dialogUI -> dialogUI.dialogId == dialogId.dialogId }
		updatePlayerDialogs(gamePlayerKey)
	}

	fun <P : Player> getDialogsForPlayer(
			game: Game<P>,
			player: P
	): UnmodifiableList<DialogUI> {
		return UnmodifiableList(playersDialogs[Pair(game.identifier, player.name)])
	}

	class DialogUI constructor(dialog: Dialog, val dialogId: String) : Dialog(dialog.title, dialog.message, dialog.form, dialog.isCancellable)

	companion object {
		private val log = LoggerFactory.getLogger(DialogManager::class.java)
	}
}
